### ISSUE TITLE
How to use this as AWS Lambda

### ISSUE THREAD

**Issue by pradeepkusingh** [apollographql/router#364]
I would like to deploy this as AWS Lambda serverless, please help any docs, example would be helpful.

**Comment by medelman17** [comment]
Hi, @pradeepkusingh! Great minds think alike, eh? As luck would have it, I've been working on a reference architecture and write-up to demonstrate using the router in a serverless function. It's still not fully baked, but watch this space and I'll make sure to ping you when it's ready.

Cheers!

**Comment by savnik** [comment]
We will also start experimenting with this so if you have any WIP we would love to see it/review and give feedback

**Comment by pradeepkusingh** [comment]
@medelman17  Have you had a chance to make this API AWS Lambda compliant?

**Comment by adikari** [comment]
any updates on this? I am currently using the nodejs version of the gateway in AWS lambda and would be lovely to start using the apollo router instead.

**Comment by pradeepkusingh** [comment]
Any update on this requests?

**Comment by jaypea** [comment]
has anyone tried a custom docker image for lambda by adding the lambda-ric wrapper around the existing docker image?

**Comment by bradens** [comment]
> has anyone tried a custom docker image for lambda by adding the lambda-ric wrapper around the existing docker image?

I don't think this works as the docker container boots an http server right?

**Comment by Tehnix** [comment]
I gave it another shot, this time by pulling in `apollo-router` as a library, and then building it myself. A bit of a similar method as above where I let the Router start its HTTP server and proxy Lambda Events to it, although this time a bit more clean since it's all kept within the same Rust program (no external binary).

You can see how exactly that is set up in this repo here https://github.com/codetalkio/the-stack/tree/d68a638bfcff7fa035f636f3abb50d3ba14548b6/ms-apollo.


I currently get the following error at the very last step, which I assume is the linking:

```bash
$ cargo lambda build # uses cargo-zigbuild
   ...
   Compiling deno_crypto v0.129.0
   Compiling opentelemetry-proto v0.3.0
   Compiling opentelemetry-otlp v0.13.0
   Compiling tikv-jemallocator v0.5.4
   Compiling ms-apollo v0.1.0 (~/GitHub/codetalkio/the-real-stack/ms-apollo)
error: linking with `~/Library/Caches/cargo-zigbuild/0.17.3/zigcc-x86_64-unknown-linux-gnu.sh` failed: exit status: 1
  |
  = note: LC_ALL="C" PATH="~/.rustup/toolchains/stable-aarch64-apple-darwin/lib/rustlib/aarch64-apple-darwi
  .....huge section of path~
-gc-sections" "-pie" "-Wl,-z,relro,-z,now" "-Wl,-O1" "-Wl,--strip-all" "-nodefaultlibs"
  = note: ld.lld: error: undefined symbol: __start_linkm2_PLUGINS
          >>> referenced by apollo_router.c57000e1db575776-cgu.00
          >>>               apollo_router-4b09e49663ee2f23.apollo_router.c57000e1db575776-cgu.00.rcgu.o:(apollo_router::configuration::_::_$LT$impl$u20$schemars..JsonSchema$u20$for$u20$apollo_router..configuration..Configuration$GT$::json_schema::heb9047cee3a0bda9) in archive ~/GitHub/codetalkio/the-real-stack/ms-apollo/target/x86_64-unknown-linux-gnu/release/deps/libapollo_router-4b09e49663ee2f23.rlib
          >>> referenced by apollo_router.c57000e1db575776-cgu.08
          >>>               apollo_router-4b09e49663ee2f23.apollo_router.c57000e1db575776-cgu.08.rcgu.o:(apollo_router::router_factory::create_plugins::_$u7b$$u7b$closure$u7d$$u7d$::h10977a494c764e3e) in archive ~/GitHub/codetalkio/the-real-stack/ms-apollo/target/x86_64-unknown-linux-gnu/release/deps/libapollo_router-4b09e49663ee2f23.rlib
          >>> referenced by apollo_router.c57000e1db575776-cgu.11
          >>>               apollo_router-4b09e49663ee2f23.apollo_router.c57000e1db575776-cgu.11.rcgu.o:(schemars::gen::SchemaGenerator::json_schema_internal::h88869beee2a5a5e8) in archive ~/GitHub/codetalkio/the-real-stack/ms-apollo/target/x86_64-unknown-linux-gnu/release/deps/libapollo_router-4b09e49663ee2f23.rlib
          >>> referenced 3 more times
          >>> the encapsulation symbol needs to be retained under --gc-sections properly; consider -z nostart-stop-gc (see https://lld.llvm.org/ELF/start-stop-gc)

          ld.lld: error: undefined symbol: __stop_linkm2_PLUGINS
          >>> referenced by apollo_router.c57000e1db575776-cgu.00
          >>>               apollo_router-4b09e49663ee2f23.apollo_router.c57000e1db575776-cgu.00.rcgu.o:(apollo_router::configuration::_::_$LT$impl$u20$schemars..JsonSchema$u20$for$u20$apollo_router..configuration..Configuration$GT$::json_schema::heb9047cee3a0bda9) in archive ~/GitHub/codetalkio/the-real-stack/ms-apollo/target/x86_64-unknown-linux-gnu/release/deps/libapollo_router-4b09e49663ee2f23.rlib
          >>> referenced by apollo_router.c57000e1db575776-cgu.08
          >>>               apollo_router-4b09e49663ee2f23.apollo_router.c57000e1db575776-cgu.08.rcgu.o:(apollo_router::router_factory::create_plugins::_$u7b$$u7b$closure$u7d$$u7d$::h10977a494c764e3e) in archive ~/GitHub/codetalkio/the-real-stack/ms-apollo/target/x86_64-unknown-linux-gnu/release/deps/libapollo_router-4b09e49663ee2f23.rlib
          >>> referenced by apollo_router.c57000e1db575776-cgu.11
          >>>               apollo_router-4b09e49663ee2f23.apollo_router.c57000e1db575776-cgu.11.rcgu.o:(schemars::gen::SchemaGenerator::json_schema_internal::h88869beee2a5a5e8) in archive ~/GitHub/codetalkio/the-real-stack/ms-apollo/target/x86_64-unknown-linux-gnu/release/deps/libapollo_router-4b09e49663ee2f23.rlib
          >>> referenced 3 more times

          ld.lld: error: undefined symbol: __start_linkme_PLUGINS
          >>> referenced by apollo_router.c57000e1db575776-cgu.00
          >>>               apollo_router-4b09e49663ee2f23.apollo_router.c57000e1db575776-cgu.00.rcgu.o:(apollo_router::configuration::_::_$LT$impl$u20$schemars..JsonSchema$u20$for$u20$apollo_router..configuration..Configuration$GT$::json_schema::heb9047cee3a0bda9) in archive ~/GitHub/codetalkio/the-real-stack/ms-apollo/target/x86_64-unknown-linux-gnu/release/deps/libapollo_router-4b09e49663ee2f23.rlib
          >>> referenced by apollo_router.c57000e1db575776-cgu.08
          >>>               apollo_router-4b09e49663ee2f23.apollo_router.c57000e1db575776-cgu.08.rcgu.o:(apollo_router::router_factory::create_plugins::_$u7b$$u7b$closure$u7d$$u7d$::h10977a494c764e3e) in archive ~/GitHub/codetalkio/the-real-stack/ms-apollo/target/x86_64-unknown-linux-gnu/release/deps/libapollo_router-4b09e49663ee2f23.rlib
          >>> referenced by apollo_router.c57000e1db575776-cgu.08
          >>>               apollo_router-4b09e49663ee2f23.apollo_router.c57000e1db575776-cgu.08.rcgu.o:(apollo_router::router_factory::create_plugins::_$u7b$$u7b$closure$u7d$$u7d$::h10977a494c764e3e) in archive ~/GitHub/codetalkio/the-real-stack/ms-apollo/target/x86_64-unknown-linux-gnu/release/deps/libapollo_router-4b09e49663ee2f23.rlib
          >>> referenced 5 more times
          >>> the encapsulation symbol needs to be retained under --gc-sections properly; consider -z nostart-stop-gc (see https://lld.llvm.org/ELF/start-stop-gc)

          ld.lld: error: undefined symbol: __stop_linkme_PLUGINS
          >>> referenced by apollo_router.c57000e1db575776-cgu.00
          >>>               apollo_router-4b09e49663ee2f23.apollo_router.c57000e1db575776-cgu.00.rcgu.o:(apollo_router::configuration::_::_$LT$impl$u20$schemars..JsonSchema$u20$for$u20$apollo_router..configuration..Configuration$GT$::json_schema::heb9047cee3a0bda9) in archive ~/GitHub/codetalkio/the-real-stack/ms-apollo/target/x86_64-unknown-linux-gnu/release/deps/libapollo_router-4b09e49663ee2f23.rlib
          >>> referenced by apollo_router.c57000e1db575776-cgu.08
          >>>               apollo_router-4b09e49663ee2f23.apollo_router.c57000e1db575776-cgu.08.rcgu.o:(apollo_router::router_factory::create_plugins::_$u7b$$u7b$closure$u7d$$u7d$::h10977a494c764e3e) in archive ~/GitHub/codetalkio/the-real-stack/ms-apollo/target/x86_64-unknown-linux-gnu/release/deps/libapollo_router-4b09e49663ee2f23.rlib
          >>> referenced by apollo_router.c57000e1db575776-cgu.11
          >>>               apollo_router-4b09e49663ee2f23.apollo_router.c57000e1db575776-cgu.11.rcgu.o:(schemars::gen::SchemaGenerator::json_schema_internal::h88869beee2a5a5e8) in archive ~/GitHub/codetalkio/the-real-stack/ms-apollo/target/x86_64-unknown-linux-gnu/release/deps/libapollo_router-4b09e49663ee2f23.rlib
          >>> referenced 3 more times

error: could not compile `ms-apollo` (bin "ms-apollo") due to previous error
```

Way out of my depth here, but if anyone has any suggestions on where to look/dig into, then I can see if I can take it further 😊

**Comment by Tehnix** [comment]
I couldn't let this go: Here's a proof-of-concept repo https://github.com/codetalkio/apollo-router-lambda.

The PoC builds the Router in an Amazon Linux 2 container so that we can actually run it in the Lambda runtime environment, and wraps it up in [cargo-lambda](https://www.cargo-lambda.info/).

TL;DR: 

- Both ways are hacky, and is built in an AL2 container
- Both have bad Cold Start times (1.3s-1.6s)
- Only one variant has acceptable Warm Start Times (49ms which
- If you want to, you can use the prebuilt binaries directly

Any PRs with improvements or ideas are welcome 😀

**Comment by abernix** [comment]
This ticket is pretty old at this point, but we've removed the Deno dependency within Router in 2.x.  We've also maintained as old glibc support as possible, with 2.28 [ref](https://github.com/apollographql/router/pull/7355).

I tried to get the reproductions above running with the latest version, but wasn't successful with the amount of effort I was able to put in.  I'm going to close this because there hasn't been much activity recently, but it's entirely conceivable that this works at this point if someone was really interested in this.

Generally though, our recommendation remains that we suggest running router in a container environment like AWS Fargate.  Happy to hear about anyone's success if they do try and get this working with 2.x though!
